client/app/bundles/course/assessment/question/programming/components/OnlineEditorCppView.jsx

```diff
@@ -0,0 +1,515 @@
+    deleteCppTestCase: PropTypes.func.isRequired,
+    updateCppNewDataFile: PropTypes.func.isRequired,
+    deleteCppNewDataFile: PropTypes.func.isRequired,
+    deleteCppExistingDataFile: PropTypes.func.isRequired,
+  }),
+  isLoading: PropTypes.bool.isRequired,
+  autograded: PropTypes.bool.isRequired,
+  intl: intlShape.isRequired,
+};
+
+const contextTypes = {
+  muiTheme: React.PropTypes.object.isRequired,
+};
+
+export function validation(data, pathOfKeysToData, intl) {
+  const errors = [];
+  const cppData = data.getIn(pathOfKeysToData);
+  if (data.getIn(['question', 'autograded'])) {
+    let testsCount = 0;
+
+    ['public', 'private', 'evaluation'].forEach((type) => {
+      const testCases = cppData.getIn(['test_cases', type]);
+      testsCount += testCases.size;
+
+      testCases.forEach((testCase, index) => {
+        const testCaseError = {};
+        let hasError = false;
+
+        if (testCase.get('expression').trim() === '') {
+          testCaseError.expression =
+            intl.formatMessage(translations.cannotBeBlankValidationError);
+          hasError = true;
+        }
+        if (testCase.get('expected').trim() === '') {
+          testCaseError.expected =
+            intl.formatMessage(translations.cannotBeBlankValidationError);
+          hasError = true;
+        }
+
+        if (hasError) {
+          errors.push({
+            path: pathOfKeysToData.concat(['test_cases', type, index, 'error']),
+            error: testCaseError,
+          });
+        }
+      });
+    });
+
+    if (testsCount === 0) {
+      errors.push({
+        path: pathOfKeysToData.concat(['test_cases', 'error']),
+        error: intl.formatMessage(translations.noTestCaseErrorAlert),
+      });
+    }
+  }
+
+  return errors;
+}
+
+class OnlineEditorCppView extends React.Component {
+
+  static getInputName(field) {
+    return `question_programming[${field}]`;
+  }
+
+  static getTestInputName(type, field) {
+    return `question_programming[test_cases][${type}][][${field}]`;
+  }
+
+  codeChangeHandler(field) {
+    return e => this.props.actions.updateCppCodeBlock(field, e);
+  }
+
+  newDataFileChangeHandler(index) {
+    return (e) => {
+      const files = e.target.files;
+      const filename = files.length === 0 ? null : files[0].name;
+      this.props.actions.updateCppNewDataFile(filename, index);
+    };
+  }
+
+  testCaseDeleteHandler(type, index) {
+    return (e) => {
+      e.preventDefault();
+
+      if (!this.props.isLoading) {
+        this.props.actions.deleteCppTestCase(type, index);
+      }
+    };
+  }
+
+  testCaseCreateHandler(type) {
+    return (e) => {
+      e.preventDefault();
+
+      if (!this.props.isLoading) {
+        this.props.actions.createCppTestCase(type);
+      }
+    };
+  }
+
+  renderExistingDataFiles = () => {
+    if (this.props.data.get('data_files').size === 0) {
+      return null;
+    }
+
+    const formatBytes = (bytes, decimals) => {
```

Comment:
> I was thinking that `OnlineEditorCppView` and `OnlineEditorPythonView` both use the same functions, perhaps it would be better to define a super class to contain these functions. And then have `OnlineEditorCppView` and `OnlineEditorPythonView` extend that super class as a form of abstraction, any additional functionalities can be overwritten or defined in the subclasses. I'm not too familiar with react, would this be a better way of handling things?

As far as I know (based on my understanding after working wiht React) there are two approaches to this: 
1) Define a general Component (like what you say, a 'super class'), then pass in props (which can be variables or functions) based on the language; or
2) Abstract out the components (eg. data file, test cases) into smaller, logic-less components to be reused. This does result in some repeated code across languages though. 

I would prefer (2) over (1), because that delegates the composability to each language, making it easier to modify should the need arise. 

However, if you have a stronger case for (1), do sound off and we can discuss. 

cc: @nicholette since you have worked a little more with React, let us know if you have any thoughts?